e2e/testcases/presync_test.go

```diff
@@ -0,0 +1,283 @@
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"testing"
+
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/policy"
+	"kpt.dev/configsync/e2e/nomostest/registryproviders"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/e2e/nomostest/testpredicates"
+	"kpt.dev/configsync/e2e/nomostest/testwatcher"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/core/k8sobjects"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reposync"
+	"kpt.dev/configsync/pkg/rootsync"
+	"kpt.dev/configsync/pkg/status"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestAddPreSyncAnnotationRepoSync(t *testing.T) {
+	repoSyncID := core.RepoSyncID(configsync.RepoSyncName, namespaceRepo)
+	nt := nomostest.New(t, nomostesting.SyncSource,
+		ntopts.RequireOCIProvider,
+		ntopts.SyncWithGitSource(repoSyncID),
+		ntopts.RepoSyncPermissions(policy.RBACAdmin(), policy.CoreAdmin()))
+	if err := nomostest.SetupPreSync(nt); err != nil {
+		nt.T.Fatal(err)
+	}
+	rootSyncGitRepo := nt.SyncSourceGitReadWriteRepository(nomostest.DefaultRootSyncID)
+	repoSyncKey := repoSyncID.ObjectKey
+	gitSource := nt.SyncSources[repoSyncID]
+
+	nt.T.Log("Create first OCI image with latest tag.")
+	bookinfoRole := k8sobjects.RoleObject(core.Name("bookinfo-admin"))
+	image, err := nt.BuildAndPushOCIImage(repoSyncKey, registryproviders.ImageInputObjects(nt.Scheme, bookinfoRole))
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Create RepoSync with first OCI image")
+	repoSyncOCI := nt.RepoSyncObjectOCI(repoSyncKey, image.OCIImageID().WithoutDigest(), "", image.Digest)
+	nt.Must(rootSyncGitRepo.Add(nomostest.StructuredNSPath(repoSyncID.Namespace, repoSyncID.Name), repoSyncOCI))
+	nt.Must(rootSyncGitRepo.CommitAndPush("Set the RepoSync to sync from OCI"))
+	nt.Must(nt.Watcher.WatchObject(kinds.RepoSyncV1Beta1(), repoSyncID.Name, namespaceRepo,
+		testwatcher.WatchPredicates(testpredicates.RepoSyncHasSourceError(status.SourceErrorCode, "no signatures found"))))
+
+	nt.T.Log("Create second image with latest tag.")
+	bookstoreSA := k8sobjects.ServiceAccountObject("bookstore-sa")
+	image1, err := nt.BuildAndPushOCIImage(repoSyncKey, registryproviders.ImageInputObjects(nt.Scheme, bookinfoRole, bookstoreSA))
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	imageURL, err := image.RemoteAddressWithDigest()
+	if err != nil {
+		nt.T.Fatal("Failed to get first image remote URL %s", err)
+	}
+	nt.T.Logf("Signing first test image %s", imageURL)
+	_, err = nt.Shell.ExecWithDebug("cosign", "sign", imageURL, "--key", nomostest.CosignPrivateKeyPath(nt), "--yes")
+	if err != nil {
+		nt.T.Fatalf("Failed to sign first test image %s", err)
+	}
+
+	nt.T.Log("Checking no signature error still exists for second image.")
```

Comment:
It's hard to tell if the `no signature error` is from the first image, or the second one. 

What about the following test order:
1. create and push an unsigned image ==> RSync is synced successfully
2. set up the webhook server ==> RSync reports a `no signature error`. This is to ensure we can verify unverified image even though the digest is the same.
3. sign the first image ==> RSync is synced successfully again. This is to ensure the verification works.
4. push another unassigned image with the same `latest` tag ==> RSync reports a `no signature error` for the new image. This is to ensure we can verify different images with floating tag.
5. other checks

Reply:
Just tried -

After step 1 the source annotations are already added so there would be no webhook request sent for the same image, thus no verification is performed. This is one of the catch: webhook can only verify later images after setup, not the current syncing one.

Reply:
I raised the issue in https://github.com/GoogleContainerTools/kpt-config-sync/pull/1452#discussion_r1802160959. We need to update the webhook server to be able to verify this scenario.

Reply:
See https://github.com/GoogleContainerTools/kpt-config-sync/pull/1452#discussion_r1813308352